webhook/webhook.go

```diff
@@ -0,0 +1,104 @@
+package webhook
+
+import (
+  "crypto/hmac"
+  "crypto/sha256"
+  "encoding/hex"
+  "net/http"
+  "fmt"
+  "errors"
+  "encoding/json"
+  "bytes"
+  "encoding/gob"
+
+  log "github.com/gophish/gophish/logger"
+)
+
+const (
+  DefaultTimeoutSeconds = 10
+  MinHttpStatusErrorCode = 400
+  SignatureHeader = "X-Gophish-Signature"
+)
+
+
+//TODO
+
+type Sender interface {
+  Send(url string, secret string, data interface{}) error
+}
+
+type defaultSender struct {
+  client *http.Client
+} 
+
+var senderInstance Sender
+func newDefaultSender() Sender {
+  sn := &defaultSender{}
+  sn.client = &http.Client{
+    Timeout: DefaultTimeoutSeconds,
+  }
+  return sn
+}
+
+func SendAll(whsInfo map[string]string, data interface{}) {
+  if senderInstance == nil {
+    senderInstance = newDefaultSender()
+  }
+  for k, v := range whsInfo {
+    go func(url string, secret string) {
+          senderInstance.Send(url, secret, data)
+       }(k, v)
+  }
+}
+
+func (ds defaultSender) Send(url string, secret string, data interface{}) error {
+  jsonData, err := json.Marshal(data)
+  if err != nil {
+    log.Error(err)
+    return err
+  }
+  req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
+  data2, err := interfaceToBytes(data)
+  if err != nil {
+    log.Error(err)
+    return err
+  }
+  signat, err := sign(secret, data2)
+  req.Header.Set(SignatureHeader, signat)
+  req.Header.Set("Content-Type", "application/json")
+  resp, err := ds.client.Do(req)
+  if err != nil {
+    log.Error(err)
+    return err
+  }
+  defer resp.Body.Close()
+
+
+  //TODO
+  if resp.StatusCode >= MinHttpStatusErrorCode {
+    errMsg := fmt.Sprintf("http status of response: %s", resp.Status)
+    log.Error(errMsg)
+    return errors.New(errMsg)
+  }
+  return nil
+}
+
+func sign(secret string, data []byte) (string, error) {
+  hash1 := hmac.New(sha256.New, []byte(secret))
+  _, err := hash1.Write(data)
+  if err != nil {
+    return "", err
+  }
+  hexStr := hex.EncodeToString(hash1.Sum(nil))
+  return hexStr, nil
+}
+
+func interfaceToBytes(data interface{}) ([]byte, error) {
```

Comment:
We should be able to remove this function since we don't need it anymore.

Reply:
We need it

<img width="753" alt="Captura de pantalla 2019-11-18 a la(s) 11 46 54" src="https://user-images.githubusercontent.com/2032888/69023327-4dab6f00-09b6-11ea-9ab0-a59e2237dc0c.png">

Reply:
Why aren’t we just signing `jsonData`?

Reply:
Because the Go library function `hmac.Write()` requires `[]byte`

Reply:
I’m not at my desk so I can’t confirm that I’m not missing something, but that should be what jsonData is, since [thats what json.Marshal returns](https://golang.org/pkg/encoding/json/#Marshal).

Reply:
Right. Fixed